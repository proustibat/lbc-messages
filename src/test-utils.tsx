import React from 'react';
import configureStore, { MockStore } from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import { RootReducersType } from './state/reducers';
import { act, fireEvent } from "@testing-library/react";

export const getMockedStore = (
  initialState: RootReducersType = { messages: { items: [], loading: false } },
) => {
  const createMockStore = configureStore([thunk]);
  const mockStore: MockStore = createMockStore(initialState);
  return mockStore;
};

export const withProvider = (WrappedComponent: React.ComponentType<any>) => (
  initialState: RootReducersType,
) => (props: any) => {
  return (
    <Provider store={getMockedStore(initialState)}>
      <WrappedComponent {...props} />
    </Provider>
  );
};

export const withRouter = (children: any) => {
  return (
    <MemoryRouter>
      <Route>{children}</Route>
    </MemoryRouter>
  );
};

export const fillForm = async (
  getByTestId: (id: string) => HTMLElement,
  { title, message }: { title: string; message: string },
  shouldSubmit: boolean = false,
) => {
  const inputTitle = getByTestId('input-title').querySelector(
    'input',
  ) as HTMLInputElement;
  const inputMessage = getByTestId('input-message').querySelector(
    'textarea',
  ) as HTMLTextAreaElement;
  const buttonSubmit = getByTestId('submit-button');

  await act(async () => {
    await fireEvent.change(inputTitle, { target: { value: title } });
    await fireEvent.change(inputMessage, { target: { value: message } });
    shouldSubmit && (await fireEvent.click(buttonSubmit));
  });
};
