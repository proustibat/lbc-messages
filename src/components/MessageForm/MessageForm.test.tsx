import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import { MessageForm, MessageFormProps } from './index';
import { fillForm } from '../../test-utils';

expect.extend({ toMatchDiffSnapshot });

const props: MessageFormProps = {
  onSubmit: jest.fn(),
};

describe('MessagesList', () => {
  afterEach(() => {
    cleanup();
    jest.resetAllMocks();
  });

  it('should render the form correctly by default', () => {
    // Given / When
    const { container } = render(<MessageForm {...props} />);

    // Then
    expect(container).toMatchSnapshot();
  });

  it('should render the form correctly wih a loader', () => {
    // Given / When
    const { container } = render(<MessageForm {...props} loading={true} />);

    // Then
    expect(container).toMatchSnapshot();
  });

  it('should enable the submit button when fields have been filled', async () => {
    // Given
    const { asFragment, getByTestId } = render(<MessageForm {...props} />);
    const firstRender = asFragment().querySelector('button');

    // When
    await fillForm(getByTestId, { title: "title", message: "message" }, false);
    const secondRender = asFragment().querySelector('button');

    // Then
    expect(firstRender).toMatchDiffSnapshot(secondRender);
  });

  it('should submit the filled data', async () => {
    // Given
    const { getByTestId } = render(<MessageForm {...props} />);

    // When
    const title = 'My awesome title';
    const message = 'My awesome message';
    await fillForm(getByTestId, { title, message }, true);

    // Then
    expect(props.onSubmit).toHaveBeenCalledTimes(1);
    expect(props.onSubmit).toHaveBeenLastCalledWith({ title, message });
  });
});
