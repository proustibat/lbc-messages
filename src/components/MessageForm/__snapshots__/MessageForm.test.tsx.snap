// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesList should enable the submit button when fields have been filled 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,10 +1,9 @@
  <button
-   class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled\\"
+   class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary\\"
    data-testid=\\"submit-button\\"
-   disabled=\\"\\"
-   tabindex=\\"-1\\"
+   tabindex=\\"0\\"
    type=\\"button\\"
  >
    <span
      class=\\"MuiButton-label\\"
    >
@@ -18,6 +17,9 @@
        <path
          d=\\"M2.01 21L23 12 2.01 3 2 10l15 2-15 2z\\"
        />
      </svg>
    </span>
+   <span
+     class=\\"MuiTouchRipple-root\\"
+   />
  </button>"
`;

exports[`MessagesList should render the form correctly by default 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-1 root_background-#fafafa root_borderRadius-4 root_boxShadow-0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12) root_padding-16"
  >
    <form
      autocomplete="off"
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiTextField-root textField_display-block textField_marginBottom-16 MuiFormControl-fullWidth"
        data-testid="input-title"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="title"
          id="title-label"
        >
          title
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="title"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                title
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root textField_display-block textField_marginBottom-16 MuiFormControl-fullWidth"
        data-testid="input-message"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="message"
          id="message-label"
        >
          message
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
            id="message"
            rows="4"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                message
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        data-testid="submit-button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          send
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root rightIcon_marginLeft-16"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
            />
          </svg>
        </span>
      </button>
    </form>
  </div>
</div>
`;

exports[`MessagesList should render the form correctly wih a loader 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-6 root_background-#fafafa root_borderRadius-4 root_boxShadow-0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12) root_padding-16"
  >
    <form
      autocomplete="off"
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiTextField-root textField_display-block textField_marginBottom-16 MuiFormControl-fullWidth"
        data-testid="input-title"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="title"
          id="title-label"
        >
          title
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="title"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-9"
            >
              <span>
                title
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root textField_display-block textField_marginBottom-16 MuiFormControl-fullWidth"
        data-testid="input-message"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="message"
          id="message-label"
        >
          message
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
            id="message"
            rows="4"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-9"
            >
              <span>
                message
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        data-testid="submit-button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          send
          <div
            class="MuiCircularProgress-root rightIcon_marginLeft-16 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
            role="progressbar"
            style="width: 20px; height: 20px;"
          >
            <svg
              class="MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </div>
        </span>
      </button>
    </form>
  </div>
</div>
`;
